### ISSUE TITLE
Python: Unknown function

### ISSUE THREAD

**Issue by lutzroeder** [lutzroeder/netron#546]
Please share a model file to reproduce this issue.

**Comment by karta1502545** [comment]
Error loading PyTorch model. Unsupported function 'torch.add_' in 'traced_resnet_model.pt'.

I face this error...

**Comment by lutzroeder** [comment]
@karta1502545 can you share the model file?

**Comment by karta1502545** [comment]
@lutzroeder Here is the link: https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py

**Comment by lutzroeder** [comment]
@karta1502545 I ran this code and seeing below. Can you share the actual model file?
<img width="1136" alt="screenshot" src="https://user-images.githubusercontent.com/438516/89378019-d171cc80-d6a7-11ea-98b8-2e037397d7de.png">

**Comment by karta1502545** [comment]
[origin_main.txt](https://github.com/lutzroeder/netron/files/5026471/origin_main.txt)

I add three line at line138-140.

The tutorial of how to use the code can be found here:
https://github.com/pytorch/examples/tree/97304e232807082c2e7b54c597615dc0ad8f6173/imagenet

**Comment by lutzroeder** [comment]
@karta1502545 can you share the traced_resnet_model.pt file?

**Comment by karta1502545** [comment]
@lutzroeder Here is my file: https://github.com/karta1502545/pth

**Comment by karta1502545** [comment]
@lutzroeder Thanks, now it works!

**Comment by clive819** [comment]
Unsupported function 'torch.group_norm'

**Comment by lutzroeder** [comment]
@clive819 can you share the model file to reproduce the issue?

**Comment by clive819** [comment]
@lutzroeder Sure! Here is the file link: https://www.dropbox.com/s/fjbp265qni7xyfs/iOS.pt?dl=0

**Comment by rajb245** [comment]
![screenshot](https://user-images.githubusercontent.com/10854480/92158048-5102d080-edf9-11ea-96a9-52ae5b6a1dd4.png)
As seen above, the error I see is:
**Unsupported function 'torch.fft' in 'traced_fft.pt'**

Here is the pt file that causes the issue (renamed to .zip extension for github upload):
[traced_fft.zip](https://github.com/lutzroeder/netron/files/5170758/traced_fft.zip)

And the following code will generate the pt file as well:
```
import torch

def forward(x, normalized=True):
    return x.fft(1, normalized)

batch_size = 4
time_steps = 2**12
real_imag = 2
x = torch.randn(batch_size,time_steps,real_imag)

traced_model = torch.jit.trace(forward,x)
assert(torch.all((traced_model(x)-forward(x))==0))
traced_model.save('traced_fft.pt')
```

**Comment by lutzroeder** [comment]
[traced_fft.zip](https://github.com/lutzroeder/netron/files/5172197/traced_fft.zip)

**Comment by myworldraj** [comment]
Unsupported function 'ops.quantized.add_relu' in 'resnet50_pertensor.pt' (one of the quantized resnet models that I am using, attached herewith
[resnet50_pertensor.zip](https://github.com/lutzroeder/netron/files/5199960/resnet50_pertensor.zip)
)

I am using the v 4.5.0 of Netron (latest available version)

**Comment by lutzroeder** [comment]
[resnet50_pertensor.zip](https://github.com/lutzroeder/netron/files/5205363/resnet50_pertensor.zip)

**Comment by myworldraj** [comment]
> [resnet50_pertensor.zip](https://github.com/lutzroeder/netron/files/5205363/resnet50_pertensor.zip)

That works. Thanks

**Comment by myworldraj** [comment]
Another unsupported function - ops.quantized.cat for the quantized inceptionv3 model that I am working with.

Attached.
[iv3_pertensor.zip](https://github.com/lutzroeder/netron/files/5208531/iv3_pertensor.zip)

**Comment by lutzroeder** [comment]
[resnet50_pertensor.zip](https://github.com/lutzroeder/netron/files/5212006/resnet50_pertensor.zip)
[iv3_pertensor.zip](https://github.com/lutzroeder/netron/files/5212009/iv3_pertensor.zip)

**Comment by rosrad** [comment]
Unsupported function 'ops.quantized.quantize_per_channel' and  'ops.quantized.conv1d_relu'  in 'test.8bit.pt' (one of the quantized CNN models), please find [https://github.com/rosrad/pth_file/blob/master/test.8bit.pth](url)

**Comment by lutzroeder** [comment]
[test.8bit.pth.zip](https://github.com/lutzroeder/netron/files/5238524/test.8bit.pth.zip)

**Comment by rosrad** [comment]
hi @lutzroeder , can I have a updated version to check the model ?

**Comment by lutzroeder** [comment]
@rosrad [DEVELOPMENT.md](https://github.com/lutzroeder/netron/blob/main/DEVELOPMENT.md)

**Comment by rajb245** [comment]
> hi @lutzroeder , can I have a updated version to check the model ?

@rosrad after the npm dependencies are installed, i just use `setup.py bdist_wheel` to make a wheel, pip install that wheel, and keep the wheel archived so i can quickly install it on other machines i have running the same setup

**Comment by lutzroeder** [comment]
[posemodel.pt.zip](https://github.com/lutzroeder/netron/files/5475350/posemodel.pt.zip)

**Comment by matthewfranglen** [comment]
Error loading PyTorch model. Unsupported function 'torch.arange' in 'ViT-B-32.pt'.

This is from the CLIP model recently released by openai and available [here](https://openaipublic.azureedge.net/clip/models/40d365715913c9da98579312b702a82c18be219cc2a73407c4526f58eba950af/ViT-B-32.pt).

**Comment by lutzroeder** [comment]
[ViT-B-32.pt.zip](https://github.com/lutzroeder/netron/files/5790183/ViT-B-32.pt.zip)

**Comment by duduheihei** [comment]
Error loading PyTorch model. Unsupported function 'opts.quantized.conv3d_prepack' in 'quant_3d.pt'.

**Comment by lutzroeder** [comment]
@duduheihei can you share the model file to reproduce the issue?

**Comment by duduheihei** [comment]
> @duduheihei can you share a model file to reproduce the issue?
You can download from this link
[quant_3d.pt](https://github.com/duduheihei/Netron/blob/master/quant_3d.pt)

**Comment by lutzroeder** [comment]
[quant_3d.pt.zip](https://github.com/lutzroeder/netron/files/5877566/quant_3d.pt.zip)

**Comment by v-goncharenko** [comment]
> **Error loading PyTorch model.**
> 
> Unsupported function 'torch.cpu' in 'pspnet.pth'.
> 
> Please provide information about this issue at https://github.com/lutzroeder/netron/issues/546.

But my model is too big of github =(

It was scripted TorchScript model with transfer to CPU (`Tensor.cpu()`)

**Comment by lutzroeder** [comment]
@v-goncharenko can you try this script to remove the weights?

```python
import os
import shutil
import sys
import tempfile
import zipfile

zip_file_path = sys.argv[1];
print(zip_file_path)
zip_dir = tempfile.mkdtemp()

with zipfile.ZipFile(zip_file_path, 'r') as zip_file:
    zip_file.extractall(zip_dir)

for folder, subfolders, files in os.walk(zip_dir):
    if folder.endswith('/data'):
        for file in files:
            path = os.path.join(folder, file)
            size = os.path.getsize(path)
            with open(path, 'w+b') as file_handle:
                file_handle.write(bytearray(size))

with zipfile.ZipFile(zip_file_path, 'w', zipfile.ZIP_DEFLATED) as zip_file:
    for folder, subfolders, files in os.walk(zip_dir):
        for file in files:
            path = os.path.join(folder, file)
            zip_file.write(path, os.path.relpath(path, zip_dir))

shutil.rmtree(zip_dir)
```

**Comment by duduheihei** [comment]
> [quant_3d.pt.zip](https://github.com/lutzroeder/netron/files/5877566/quant_3d.pt.zip)

Thanks to support this model. However, the scale and zero_point of the quantized  parameters are not correct. Could you fix this problem?

**Comment by lutzroeder** [comment]
@duduheihei can you provide actual repro steps. What are the expected and actual values. Can you provide a fix?

**Comment by duduheihei** [comment]
> @duduheihei can you provide actual repro steps, expected and actual values or submit a fix?
[quant_param_converted.pt](https://github.com/duduheihei/Netron/blob/master/quant_param_converted.pt) stores correct parameters.
[quant_3d.pt](https://github.com/duduheihei/Netron/blob/master/quant_3d.pt) is the serilized torch script model.
I compare the parameters in this two files and find out only scales value are not same. Because I update a new quant_3d.pt model, you should download these two files again to check it.

**Comment by v-goncharenko** [comment]
> @v-goncharenko can you try this script to zero the weights which results in a much smaller file?

@lutzroeder yes, I did that - [pspnet_light.pth.zip](https://github.com/lutzroeder/netron/files/5919067/pspnet_light.pth.zip).

I would suggest not to use the same filename for writing back to a new lightweight file because it could damage original data (some folks don't make backups, you know) =)

Hope this helps

**Comment by nlgranger** [comment]
Does Netron have a fallback to generate a generic node for unsupported functions?

**Comment by lutzroeder** [comment]
@nlgranger what's an example of such a function and what indicates that a node needs to be generated for that function?

**Comment by duduheihei** [comment]
> @duduheihei can you provide actual repro steps, expected and actual values or submit a fix?

I tried the latest 4.7.6 version. However, the problem is still there. The "output_scale" showed by netron are different from that in "quant_param_converted.pt" I provided.

**Comment by lutzroeder** [comment]
@duduheihei can you help investigate this issue and submit a fix.

**Comment by duduheihei** [comment]
> @duduheihei can you help investigate this issue and submit a fix.

I really hope I could help to sovle the problem, However I have not read the source code of Netron. So it may be very difficult for me to fix it. I will try to read the code.

**Comment by duduheihei** [comment]
> @duduheihei can you help investigate this issue and submit a fix.

Hi, I opened [quant_3d.pt](https://github.com/duduheihei/Netron/blob/master/quant_3d.pt) and [quant_param_converted.pt](https://github.com/duduheihei/Netron/blob/master/quant_param_converted.pt)  with 4.7.6 version. I compare the datatype of scale in both file, and find out that the datatypes are either float32 or float64 in the first file, however always float32 in the later file. So I guess there may be something wrong with datatype parse. Can you check it?

**Comment by abdullah-al-masud** [comment]
I am facing an error showing "unsupported function torch.softmax" 
I attached the script file here.
[torch_script_model_dot15.zip](https://github.com/lutzroeder/netron/files/6307552/torch_script_model_dot15.zip)

**Comment by lutzroeder** [comment]
[torch_script_model_dot15.pt.zip](https://github.com/lutzroeder/netron/files/6308295/torch_script_model_dot15.pt.zip)

**Comment by jokerspy** [comment]
Error loading PyTorch model. Unsupported function 'torch.__rshift__' in 'test.pt'.
I met above error with netron 4.6.0 and torch 1.5.1. Minimal example to reproduce is provided as below. Thanks.

```python
import torch
import torch.nn as nn
import torch.nn.init as init

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()

    def forward(self, x):
        return x >> 1

if __name__ == '__main__':
    i = torch.randn(1, 1, 3, 3)
    model = Net()
    model.eval()

    with torch.no_grad():
        traced_script_module = torch.jit.trace(model, i)
        traced_script_module.save("test.pt")
```

**Comment by allenhsu6** [comment]
Unsupported function 'torch.cat' 

the pt file link :  

https://lyft-l5-datasets-public.s3-us-west-2.amazonaws.com/models/urban_driver/BPTT.pt

**Comment by lutzroeder** [comment]
[BPTT.pt.zip](https://github.com/lutzroeder/netron/files/8712572/BPTT.pt.zip)

**Comment by jinfagang** [comment]
Got a simillar issue:

<img width="426" alt="image" src="https://user-images.githubusercontent.com/21303438/182775155-8107a4c5-b504-4c77-86e4-4cccd921cf4a.png">

**Comment by lutzroeder** [comment]
@jinfagang can you share a model file to reproduce the issue.

**Comment by jinfagang** [comment]
@lutzroeder Hi, the model is torchscript which is very huge (499M), but the model is traced from here (torchscript): https://github.com/jinfagang/wenet

**Comment by kaintels** [comment]
Hello, Thank you for sharing great work.

I found error message related to 'fft_fft'.

Python Error Unknown function 'torch.fft_fft'.
Minimal example to reproduce is provided as below. Thanks.

python : 3.8
netron : 6.3.8
pytorch : 1.12.1

```python

import torch
import torch.nn as nn

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()

    def forward(self, x):
        return torch.fft.fft(x).real

if __name__ == '__main__':
    inputs = torch.randn(1, 1)
    model = Net()
    model.eval()

    with torch.no_grad():
        traced_script_module = torch.jit.trace(model, inputs)
        traced_script_module.save("test.pt")
```

**Comment by simonrouse9461** [comment]
I'm trying to visualize PyG models. However, it seems that it's not supported yet.

Here's a code snippet using an `NNConv` layer from PyG:
```python
from torch import nn, hit
import torch_geometric as pyg

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = pyg.nn.NNConv(8, 8, nn=nn.Linear(2, 64)).jittable()

    def forward(self, x, edge_index, edge_feat):
        return self.conv(x=x, edge_index=edge_index, edge_attr=edge_feat)

jit.save(jit.script(Model()), "pyg_model.pt")
```

Then, opening "pyg_model.pt" will give the following error:
```
Unknown function 'ops.prim.is_sparse'.
```

Please add support to PyG. Really appreciate your efforts!

**Comment by lutzroeder** [comment]
[pyg_model.zip](https://github.com/lutzroeder/netron/files/10369483/pyg_model.zip)

**Comment by pbsds** [comment]
`Python Error Unknown function 'torch.take_along_dim'.`

```python
import torch

class Test(torch.nn.Module):
    def forward(self, x):
        indices = x.argmin(dim=-2, keepdim=True)
        x = x.take_along_dim(indices, -2).squeeze(-2)
        return x.sum()

torch.jit.trace(
    Test(),
    (torch.empty(70, 4, 3)),
).save("foobar.pt")
```

[foobar.zip](https://github.com/lutzroeder/netron/files/10465688/foobar.zip)

**Comment by edisonlee55** [comment]
`Unknown function 'ops.prepacked.linear_clamp_prepack'.`

```python
import torch
import timm
from torch.utils.mobile_optimizer import optimize_for_mobile

model = timm.create_model('mobilenetv3_large_100', pretrained=True)
model.eval()
scripted_model = torch.jit.script(model)
opt_model = optimize_for_mobile(scripted_model)
opt_model._save_for_lite_interpreter("model.ptl")
```

[model.zip](https://github.com/lutzroeder/netron/files/11130441/model.zip)

**Comment by wellhowtosay** [comment]
Unknown function 'torch.sqrt' in 'model.pt'.

**Comment by lutzroeder** [comment]
@wellhowtosay can you share the model file to reproduce this issue.

**Comment by Dawars** [comment]
Unknown function 'torch._autocast_to_reduced_precision'.


Comes from U2 net wrapped in a PyTorch Lightning model and trained with 16-mixed precision.

Exported via tracing and 
```model_inference = torch.jit.optimize_for_inference(traced_model)```

[u2net.zip](https://github.com/lutzroeder/netron/files/11463805/u2net.zip)

**Comment by liquanfeng** [comment]
Unknown function 'torch.l1_loss'.
model too big to upload.

**Comment by andreaslam** [comment]
![Screenshot 2023-08-05 at 1 30 01 PM](https://github.com/lutzroeder/netron/assets/81166794/662a8434-2f9f-4c57-b3b3-d1f1444b5ad9)
It seems like `torch.pow` is not recognised on Netron as well. I'm using a `.pt` file but the `.onnx` version of the file works.

**Comment by lutzroeder** [comment]
@andreaslam can you share a model file to reproduce the issue?

**Comment by andreaslam** [comment]
@lutzroeder sure! let me attach the .pt file
Uploading chess_16x128_gen3634.pt.zip…]()

**Comment by Railcalibur** [comment]
Work with PyG: Unsupported function'ops.torch_scatter.scatter_max'.
Please add support to PyG. Really appreciate your efforts!

**Comment by lutzroeder** [comment]
@Railcalibur can you share a sample file to reproduce the issue.

**Comment by sandeep3sai** [comment]
@lutzroeder i get the unknown function error when trying to load a torch script file
![image](https://github.com/lutzroeder/netron/assets/53262588/e62b6a66-07ac-46fc-baa1-e0fd480a045c)

The model was generated using the scripts from https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/SpeechSynthesis/Tacotron2/exports/export_tacotron2_ts.py

**Comment by lutzroeder** [comment]
@sandeep3sai can you share a model file to reproduce the issue.

**Comment by nastyakor01** [comment]
Hello! Faced this problem:

Unknown function: 'torch.split'.

Netron version: 7.3.1
File causing the issue: ecapa2.pt loaded from https://huggingface.co/Jenthe/ECAPA2/tree/main

**Comment by zhenzhou12** [comment]
![image](https://github.com/lutzroeder/netron/assets/167603195/cb7cd09f-4ffb-4d24-95fb-89f11e6b2221)
My model file is in .pkl format, how can I modify it?

**Comment by lutzroeder** [comment]
@zhenzhou12 can you share a model file to reproduce the issue.

**Comment by aryan1165** [comment]
Unknown function torch.strip. 
My model file :- https://drive.google.com/file/d/1j4IdX1A6ucJ8RyKomJrsRjJKk64z2SM3/view?usp=drive_link

there are two models, both are not working

**Comment by vikramambrose** [comment]
I'm seeing  `Unknown function 'ops.prim.ConstantChunk'.` on v7.8.3 when opening a python traced module.

My reproducer is too big to upload here. Is there somewhere I can upload the model?

**Comment by lutzroeder** [comment]
@vikramambrose is it possible to create a smaller file using this operator or share it via Dropbox or Google Drive?

**Comment by lutzroeder** [comment]
This issue is a catch all for files including unknown Python functions.

Please share a model file to reproduce this issue and include additional context how the file was created.

**Comment by Anyway1223** [comment]
Unknown function 'torch.rms_norm'.
